doc/sphinx/language/extensions/compil-steps.rst

```diff
@@ -0,0 +1,154 @@
+<Scopes>` mechanism. For instance :n:`(1 + 2 * 2)%N` can be
+the above natural number expression while :n:`(1 + 2 * 2)%Z` can be
+an expression with integers.
+
+In order to take the best part of all these preprocessing mechanisms,
+one needs a basic understanding of the multiple steps needed to
+transform an inputted term (possibly with notations) into the valid
+Gallina term which Rocq will ultimately use internally. Terms given as input
+to Rocq go through several successive steps:
+
+* First, :term:`lexing`, then :term:`parsing`, are performed as part
+  of any command processing, as described above.
+
+* During :gdef:`internalization` a number of things are resolved. This
+  includes :ref:`name resolution <qualified-names>`, :term:`notation
+  interpretation` and introduction of :term:`holes <hole>` for :ref:`implicit
+  arguments <ImplicitArguments>`.
+  :gdef:`Notation interpretation <notation interpretation>`
+  translates each syntactic element to a term,
+  for instance :n:`1` can be interpreted as the
+  natural number :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`,
+  then :n:`2 * 2` as :n:`Nat.mul (S (S O)) (S (S O))` and finally our
+  whole term as :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`. The
+  same expression can be given multiple interpretations in various
+  contexts thanks to :ref:`Scopes`.
+
+* Finally, :gdef:`type inference`, can use the various mechanisms described in
+  this section to fill gaps (for instance with :ref:`canonical structures
+  <canonicalstructures>` or :ref:`typeclasses`) or fix terms (for
+  instance with :ref:`coercions <coercions>`) to obtain fully detailed terms in
+  the :ref:`Core Language <core-language>`.
+
+For each term, Rocq performs these steps *successively* and
+*independently*.
+Then, the result goes through the type checking phases discussed in
+:ref:`previous chapter <core-language>`.
+None of the steps has any impact on the previous ones.
+In particular, no typecheking is involved during parsing or
+internalization. Also note that none of the feature resolved during
+these phases, like unqualified names, implicit arguments or notations,
+remains during the later elaboration and type checking phases.
+
+.. note::
+
+   The :term:`type inference` phase together with, all or part of, the
+   previous steps is sometimes called elaboration in the literature.
+
+.. example:: Simple interleaving of intern and type inference phases
+
+   The command :g:`Definition foo : T := body.` has a trivial
+   :term:`synterp` phase. Indeed, it doesn't influence any further
+   parsing. Its :term:`interp` phase will internalize :g:`T` and infer
+   types in it, then it will internalize :g:`body` and infer types in
+   it, using :g:`T` as expected type. Note that the result of type
+   inference in :g:`T` matters in internalization of :g:`body`, for
+   instance for selecting notation scopes. Finally, the resulting
+   :g:`foo : T := body` is sent to the kernel.
```

Comment:
Is it possible to have a  concrete example instead?
For instance `Definition foo : list _ := cons 4 (cons 2 nil).`

Reply:
(But I would not block this PR for this, it can come later).

Reply:
Well, a more concrete example would have to be a bit more involved to illustrate the previous sentence (in your example `Definition foo : list _ := cons 4 (cons 2 nil).` and `Definition foo := cons 4 (cons 2 nil).` are equivalent). I'd rather not have a too complicated example.